contextual_lenses/pfam_family_classification/pfam_one_hot_lenses_parallel.py

```diff
@@ -0,0 +1,223 @@
+
+  one_hot_inds = list(df['one_hot_inds'].values)
+  one_hot_inds = tf.data.Dataset.from_tensor_slices(one_hot_inds)
+
+  indexes = df['index'].values
+  indexes = tf.data.Dataset.from_tensor_slices(indexes)
+
+  batches = tf.data.Dataset.zip((one_hot_inds, indexes)).shuffle(buffer_size=buffer_size, seed=seed)
+  batches = batches.repeat(epochs).batch(batch_size=batch_size, drop_remainder=drop_remainder).as_numpy_iterator()
+
+  return batches
+
+
+"""Model evaluation."""
+def evaluate(predict_fn, test_data, true_indexes, title, loss_fn_kwargs):
+
+  pred_indexes = []
+  cross_entropy = 0.
+  
+  for batch in iter(test_data):
+
+    X, Y = batch
+
+    Y_hat = predict_fn(X)
+
+    cross_entropy += cross_entropy_loss(Y, Y_hat, **loss_fn_kwargs)
+
+    preds = jnp.argmax(jax.nn.log_softmax(Y_hat), axis=1)
+    for pred in preds:
+      pred_indexes.append(pred)
+  pred_indexes = np.array(pred_indexes)
+  
+  acc = sklearn.metrics.accuracy_score(true_indexes, pred_indexes)
+  
+  results = {
+      'title': title,
+      'cross_entropy': cross_entropy,
+      'accuracy': acc,
+  }
+
+  return results, preds
+
+
+"""Experiments."""
+epochs = 100
+train_batches = create_data_iterator(train_df, batch_size=128*8, epochs=epochs, drop_remainder=True)
+test_batches = create_data_iterator(test_df, batch_size=512, buffer_size=1)
+lr = 1e-3
+wd = 0.
+encoder_fn = cnn_one_hot_encoder
+encoder_fn_kwargs = {
+    'n_layers': 1,
+    'n_features': [512],
+    'n_kernel_sizes': [12]
+}
+reduce_fn = max_pool
+reduce_fn_kwargs = {
+    
+}
+loss_fn_kwargs = {
+  'num_classes': num_families
+}
+cnn_max_pool_model = create_representation_model(encoder_fn=encoder_fn,
+                                                 encoder_fn_kwargs=encoder_fn_kwargs,
+                                                 reduce_fn=reduce_fn,
+                                                 reduce_fn_kwargs=reduce_fn_kwargs,
+                                                 num_categories=num_categories,
+                                                 output_features=num_families)
+
+cnn_max_pool_optimizer = p_train(model=cnn_max_pool_model,
+                               train_data=train_batches,
+                               loss_fn=cross_entropy_loss,
+                               loss_fn_kwargs=loss_fn_kwargs,
+                               learning_rate=lr,
+                               weight_decay=wd,
+                               restore_dir=restore_dir,
+                               save_dir=save_dir)
+
+del cnn_max_pool_model
+
+cnn_max_pool_results, cnn_max_pool_preds = evaluate(predict_fn=cnn_max_pool_optimizer.target,
+                                                    test_data=test_batches,
+                                                    true_indexes=test_indexes,
+                                                    title='CNN + Max Pool',
+                                                    loss_fn_kwargs=loss_fn_kwargs)
+
+print(cnn_max_pool_results)
```

Comment:
instead of writing a proper unit test for the model saving/loading, at the very least you could do something at the bottom of this file where you load a new model from disk, compute the eval metrics, and make sure they match cnn_max_pool_results